include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,275 @@
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemGetBuffer(
+    TRITONCACHE_CacheEntryItem* item, size_t index, void** base,
+    size_t* byte_size);
+
+///
+/// The following functions can be implemented by a cache. Functions
+/// indicated as required must be implemented or the cache will fail
+/// to load.
+///
+
+/// Create a new cache object.
+///
+/// This function is required to be implemented by the cache.
+///
+/// The caller takes ownership of the
+/// TRITONCACHE_Cache object and must call
+/// TRITONCACHE_CacheDelete to release the object.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache Returns the new cache object.
+/// \param config The config options to initialize the cache with.
+///               This will be passed as-is to the cache implementation, so
+///               the expected format and parsing is up to the cache as well.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheNew(
+    TRITONCACHE_Cache** cache, const char* config);
+
+/// Delete a cache object.
+///
+/// This function is required to be implemented by the cache.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache The cache object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheDelete(
+    TRITONCACHE_Cache* cache);
+
+/// Inserts entry into cache at specified key, unless key already exists.
+///
+/// This function is required to be implemented by the cache.
+///
+/// This API is implemented by the user-provided cache implementation,
+/// so specific details will be found within the cache implementation.
+///
+/// \param cache The object that is used to communicate with the cache
+///              implementation through a shared library.
+/// \param key The key used to access the cache. Generally, this is some
+///            unique value or hash representing the entry to avoid collisions.
+/// \param entry The entry to be inserted into the cache.
+/// \return a TRITONSERVER_Error indicating success or failure.
+///         Specific errors will be up the cache implementation, but general
+///         error best practices that should be followed are as follows:
+///         - TRITONSERVER_ERROR_INVALID_ARG
+///           - bad argument passed, nullptr, etc.
+///         - TRITONSERVER_ERROR_ALREADY_EXISTS
+///           - key already exists and will not be inserted again
+///         - TRITONSERVER_ERROR_INTERNAL
+///           - internal errors
+///         - nullptr
+///           - success
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheInsert(
+    TRITONCACHE_Cache* cache, const char* key, TRITONCACHE_CacheEntry* entry);
```

Comment:
I discussed this with the Redis team and they did not see value in abstracting the keys. The argument was that strings are flexible and ubiquitously used by the majority share of caches. When they're not used, they're still castable/convertible to other types fairly easily if needed. So I decided to go with the "make the common case fast" approach for the workflow.

```
# some cache test or application
cache->Insert("mykey1", obj)
cache->Lookup("mykey1")
```
vs. having to do this to setup a key for each access
```
# some cache test or application
TRITONCACHE_CacheKey* key = nullptr;
TRITONCACHE_CacheKeyNew(&key);
cache->Insert(key, obj)
TRITONCACHE_CacheKeyDelete(key);
```

Reply:
>  I think const char* key would prevent us from having key names that require `\0` in the middle of the key.
This seems like a reasonable thing for us to not support in exchange for the ease-of-use it offers. 

If we really wanted to support that, I believe including a `size_t byte_size` along with it could support that use case if copied correctly.

Reply:
I think having `byte_size` is a good enhancement. If I'm understanding correctly, there could be scenarios where the cache key is a result of a hash that contains `\0` in it. The cache implementer would need to have other workarounds to avoid this issue.